modules/avcodec/decode.c

```diff
@@ -251,42 +259,44 @@ static int ffdecode(struct viddec_state *st, struct vidframe *frame,
 
 		if (hw_frame) {
 			/* retrieve data from GPU to CPU */
-			ret = av_hwframe_transfer_data(st->pict, hw_frame, 0);
+			ret = av_hwframe_transfer_data(sw_frame, hw_frame, 0);
 			if (ret < 0) {
 				warning("avcodec: decode: Error transferring"
 					" the data to system memory\n");
 				goto out;
 			}
 
-			st->pict->key_frame = hw_frame->key_frame;
+			sw_frame->key_frame = hw_frame->key_frame;
 		}
 
-		frame->fmt = avpixfmt_to_vidfmt(st->pict->format);
+		frame->fmt = avpixfmt_to_vidfmt(sw_frame->format);
 		if (frame->fmt == (enum vidfmt)-1) {
 			warning("avcodec: decode: bad pixel format"
 				" (%i) (%s)\n",
-				st->pict->format,
-				av_get_pix_fmt_name(st->pict->format));
+				sw_frame->format,
+				av_get_pix_fmt_name(sw_frame->format));
 			goto out;
 		}
 
 		for (i=0; i<4; i++) {
-			frame->data[i]     = st->pict->data[i];
-			frame->linesize[i] = st->pict->linesize[i];
+			frame->data[i]     = sw_frame->data[i];
+			frame->linesize[i] = sw_frame->linesize[i];
```

Comment:
please note that the frame memory has to be available after this function has returned,
because the vidframe is point to those data buffers ...

Reply:
`av_free` should only `free` the allocated AVFrame not the data itself I think.